File: src/algorithm/leetcode/Problem208.java
package algorithm.leetcode;

import java.util.HashMap;

public class Problem208 {

    class TrieNode {
        public String val;
        public HashMap<String, TrieNode> nextMap;

        public TrieNode(String val) {
            this.val = val;
            nextMap = new HashMap<>();
        }
    }

    TrieNode root;

    /** Initialize your data structure here. */
    public Problem208() {
        root = new TrieNode("@");
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        StringBuilder sb = new StringBuilder(word);
        sb.append("&");
        String tmpWord = sb.toString();
        TrieNode cur = root;
        for (int i = 0; i < tmpWord.length(); i++) {
            String charac = tmpWord.substring(i, i + 1);
            if (cur.nextMap.get(charac) == null) {
                TrieNode newNode = new TrieNode(charac);
                cur.nextMap.put(charac, newNode);
            }
            cur = cur.nextMap.get(charac);
        }
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        StringBuilder sb = new StringBuilder(word);
        sb.append("&");
        String tmpWord = sb.toString();
        TrieNode cur = root;
        for (int i = 0; i < tmpWord.length(); i++) {
            String charac = tmpWord.substring(i, i + 1);
            if (cur.nextMap.get(charac) == null)
                return false;
            cur = cur.nextMap.get(charac);
        }
        return true;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for (int i = 0; i < prefix.length(); i++) {
            String charac = prefix.substring(i, i + 1);
            if (cur.nextMap.get(charac) == null)
                return false;
            cur = cur.nextMap.get(charac);
        }
        return true;
    }

    public static void main(String[] args) {
        Problem208 trie = new Problem208();
        trie.insert("apple");
        trie.insert("beer");
        System.out.println(trie.search("app"));
    }
}

File: src/algorithm_cpp/problem208.cpp
struct TrieNode {
    char val;
    vector<TrieNode*> nexts;
    TrieNode(char v) {
	val = v;
    }
};

class Trie {
private:
    TrieNode *root;

public:
    /** Initialize your data structure here. */
    Trie() {
	root = new TrieNode(0);
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
	word += '\0';
	TrieNode *curr = root;
	for (int i = 0; i < word.size(); i++) {
	    char ch = word.at(i);
	    bool flag = false;
	    for (int j = 0; j < curr->nexts.size(); j++) {
		if (curr->nexts[j]->val == ch) {
		    flag = true;
		    curr = curr->nexts[j];
		    break;
		}
	    }
	    if (flag == false) {
		TrieNode *tmp = new TrieNode(ch);
		curr->nexts.push_back(tmp);
		curr = tmp;
	    }
	}	   
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
	word += '\0';
	return startsWith(word);
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
	TrieNode *curr = root;
	for (int i = 0; i < prefix.size(); i++) {
	    char ch = prefix.at(i);
	    bool flag = false;
	    for (int j = 0; j < curr->nexts.size(); j++) {
		if (curr->nexts[j]->val == ch) {
		    flag = true;
		    curr = curr->nexts[j];
		    break;
		}
	    }
	    if (flag == false) {
		return false;
	    }
	}
	return true;
    }
};
